logicalplan/distribute_test.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"math"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/efficientgo/core/testutil"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/thanos-community/promql-engine/api"
+)
+
+func TestDistributedExecution(t *testing.T) {
+	cases := []struct {
+		name     string
+		expr     string
+		expected string
+	}{
+		{
+			name: "sum-rate",
+			expr: `sum by (pod) (rate(http_requests_total[5m]))`,
+			expected: `
+sum by (pod) (
+  dedup(coalesce(
+    remote(sum by (pod, region) (rate(http_requests_total[5m]))),
+    remote(sum by (pod, region) (rate(http_requests_total[5m])))
+  ))
+)`,
+		},
+		{
+			name: "sum-rate without labels preserves engine labels",
+			expr: `sum without (pod, region) (rate(http_requests_total[5m]))`,
+			expected: `
+sum without (pod, region) (
+  dedup(coalesce(
+    remote(sum without (pod) (rate(http_requests_total[5m]))),
+    remote(sum without (pod) (rate(http_requests_total[5m])))
+  ))
+)`,
+		},
+		{
+			name: "avg",
+			expr: `avg by (pod) (http_requests_total)`,
+			expected: `
+avg by (pod) (
+  dedup(coalesce(
+    remote(http_requests_total),
+    remote(http_requests_total)
+  ))
+)`,
+		},
+		{
+			name: "two-level aggregation",
+			expr: `max by (pod) (sum by (pod) (http_requests_total))`,
+			expected: `
+max by (pod) (
+  sum by (pod) ( 
+    dedup(coalesce(
+      remote(sum by (pod, region) (http_requests_total)),
+      remote(sum by (pod, region) (http_requests_total))
+    ))
+  )
+)`,
+		},
+		{
+			name: "aggregation of binary expression",
+			expr: `max by (pod) (metric_a / metric_b)`,
+			expected: `
+max by (pod) (
+  dedup(coalesce(remote(metric_a), remote(metric_a))) 
+  / 
+  dedup(coalesce(remote(metric_b), remote(metric_b)))
+)
+`,
+		},
+		{
+			name: "unsupported aggregation in the operand path",
+			expr: `max by (pod) (sort(avg(http_requests_total)))`,
+			expected: `
+max by (pod) (sort(avg(
+  dedup(coalesce(
+    remote(http_requests_total),
+    remote(http_requests_total)
+  ))
+)))`,
+		},
+		{
+			name: "binary operation in the operand path",
+			expr: `max by (pod) (sort(metric_a / metric_b))`,
+			expected: `
+max by (pod) (sort(
+  dedup(coalesce(remote(metric_a), remote(metric_a))) 
```

Comment:
In the future, it would be nice to add `Name() string` or a similar method to `api.RemoteEngine` so that we could use it in Explain() and other places.